### DISCUSSION TITLE
How to reduce the memory required to run？

### DISCUSSION THREAD

**Discussion by DYLDYLDYL** [idaholab/moose#22618] [Q&A General] [answered]
Hi！

My laptop's memory is 16G, and the program to run needs 17G.  I want to test it on the laptop first, and then run it on the server.  Is there any way to reduce the memory required to run？

Thanks in advance for your help!

**Comment by GiudGiud** [comment, marked as answer]
Hello

If you are running in parallel, switch to serial to reduce memory.
You should remove any auxiliary variable that you dont really need as they cost memory.
If you can, the easiest solution is to run with a coarser mesh.
Another thing to improve is the preconditioning. What are you using right now?

Guillaume

**Reply by DYLDYLDYL** [reply]
Thanks！
The distribution of memory is as follows
```
  Finished Reading Mesh                                                       [  0.34 s] [  132 MB]
  Finished Setting Mesh                                                         [  0.61 s] [  244 MB]
  Setting Up Displaced Mesh
  Finished Preparing Mesh                                                    [  0.64 s] [  108 MB]
  Finished Setting Up Displaced Mesh                                  [  0.64 s] [  108 MB]
  Finished Initializing Equation Systems                                [  2.02 s] [  751 MB]
  Finished Initializing Displaced Equation System                 [  0.49 s] [  215 MB]
  Finished Initializing                                                              [  4.16 s] [  995 MB]
  Finished Setting Up                                                             [  6.33 s] [ 1511 MB]
  Finished Projecting Initial Solutions                                     [  6.18 s] [    0 MB]
  Finished Computing Initial Material Values                          [  8.42 s] [ 4204 MB]
  Finished Setting Up Materials                                              [  8.42 s] [ 4204 MB]
  Updating Geometric Search
  Updating Displaced GeometricSearch
  Finding Nearest Nodes                                                                [  6.81 s] [  277 MB]
  Finished Updating Displaced GeometricSearch                          [  6.81 s] [  277 MB]
  Finished Updating Geometric Search                                          [  6.81 s] [  277 MB]
  Reinitializing Because of Geometric Search Objects..                 [ 16.63 s] [  361 MB]
```

I have only two auxiliary variables coupled time. 
My mesh has only 27w elems.
Precondition:
```
  petsc_options_iname = '-pc_type -pc_factor_shift_type -pc_factor_mat_solver_type'
  petsc_options_value = 'lu NONZERO superlu_dist'
  line_search = none
```
And I wonder whether the simulation results and convergence ability are affected by not using preprocessing.
Thanks in advance for your help!

**Reply by GiudGiud** [reply]
Preconditioning is usually necessary. 
You can try some less-memory consuming forms, like Jacobi or block jacobi. 
You can try to use non-full Jacobian preconditioning with PJFNK that will also consume less. 

I see 4GB in materials. I imagine they are stateful materials because we save the initial value. 
Are you outputting the material properties as well? That costs memory

**Reply by DYLDYLDYL** [reply]
Thanks!
I don't need to output the material properties,  and I wonder how to set it. 
My [output] is only exodus = true.

**Reply by GiudGiud** [reply]
to output material properties you add output = in the Materials block. 

If all the other options dont work, I would reduce the mesh for this local computation

**Reply by DYLDYLDYL** [reply]
Thanks! I will try it.

**Comment by friedmud** [comment]
Did you get this going?  Most-likely you were running out of memory due to using LU.  As @GiudGiud mentioned you might do better using `asm` or `bjacobi`.

Another thing to look into is running with distributed mesh: https://mooseframework.inl.gov/syntax/Mesh/splitting.html

**Reply by DYLDYLDYL** [reply]
Thanks！I found that the excessive memory usage was caused by my unreasonable mesh, and it has been resolved now！
